fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/domain/transactionprocessor/impl/AdvancedPaymentScheduleTransactionProcessor.java

```diff
@@ -1527,6 +1507,74 @@ private Money processAllocationsHorizontally(LoanTransaction loanTransaction, Tr
+            Set<LoanCharge> inAdvanceInstallmentCharges) {
+        Money paidPortion;
+        ProgressiveLoanInterestScheduleModel model = ctx.getModel();
+        LocalDate payDate = loanTransaction.getTransactionDate();
+        if (DueType.IN_ADVANCE.equals(paymentAllocationType.getDueType())) {
+            payDate = calculateNewPayDateInCaseOfInAdvancePayment(loanTransaction, inAdvanceInstallment);
+            updateRepaymentPeriodBalances(paymentAllocationType, inAdvanceInstallment, model, payDate);
+        }
+
+        paidPortion = processPaymentAllocation(paymentAllocationType, inAdvanceInstallment, loanTransaction, transactionAmountUnprocessed,
+                loanTransactionToRepaymentScheduleMapping, inAdvanceInstallmentCharges, balances,
+                LoanRepaymentScheduleInstallment.PaymentAction.PAY);
+
+        if (PRINCIPAL.equals(paymentAllocationType.getAllocationType())) {
+            emiCalculator.payPrincipal(model, inAdvanceInstallment.getDueDate(), payDate, paidPortion);
+            updateRepaymentPeriods(loanTransaction, ctx, model);
+        } else if (INTEREST.equals(paymentAllocationType.getAllocationType())) {
+            emiCalculator.payInterest(model, inAdvanceInstallment.getDueDate(), payDate, paidPortion);
+            updateRepaymentPeriods(loanTransaction, ctx, model);
+        }
+        return paidPortion;
+    }
+
+    private static void updateRepaymentPeriods(LoanTransaction loanTransaction, ProgressiveTransactionCtx ctx,
+            ProgressiveLoanInterestScheduleModel model) {
+        model.repaymentPeriods().forEach(rm -> {
+            LoanRepaymentScheduleInstallment installment = ctx.getInstallments().stream()
+                    .filter(ri -> ri.getDueDate().equals(rm.getDueDate()) && !ri.isDownPayment()).findFirst().orElse(null);
+            if (installment != null) {
+                installment.updatePrincipal(rm.getDuePrincipal().getAmount());
+                installment.updateInterestCharged(rm.getDueInterest().getAmount());
+                installment.updateObligationsMet(ctx.getCurrency(), loanTransaction.getTransactionDate());
+            }
+        });
+    }
+
+    private void updateRepaymentPeriodBalances(PaymentAllocationType paymentAllocationType,
+            LoanRepaymentScheduleInstallment inAdvanceInstallment, ProgressiveLoanInterestScheduleModel model, LocalDate payDate) {
+        PayableDetails payableDetails = emiCalculator.getPayableDetails(model, inAdvanceInstallment.getDueDate(), payDate);
+
+        switch (paymentAllocationType) {
+            case IN_ADVANCE_INTEREST -> inAdvanceInstallment.updateInterestCharged(payableDetails.getPayableInterest().getAmount());
+            case IN_ADVANCE_PRINCIPAL -> inAdvanceInstallment.updatePrincipal(payableDetails.getPayablePrincipal().getAmount());
+            default -> {
+            }
+        }
+    }
+
+    private static LocalDate calculateNewPayDateInCaseOfInAdvancePayment(LoanTransaction loanTransaction,
```

Comment:
Same here, probably shouldn't be static.

Reply:
I didn't see this, thanks.